src/OrchardCore.Modules/OrchardCore.Templates/Controllers/TemplateController.cs

```diff
@@ -352,13 +352,17 @@ public async Task<ActionResult> ListPost(ContentOptions options, IEnumerable<str
         {
             return Forbid();
         }
+        
+        // Prevent multiple enumeration.
+        var itemIdList = itemIds as IList<string> ?? itemIds?.ToList() ?? new List<string>();
```

Comment:
I am not sure this is needed. Can you explain how this prevents multiple enumerations?

Reply:
[Here](https://github.com/OrchardCMS/OrchardCore/pull/16623/files/3fbb66b6aef1ee3b8ed920741c6bf2162bc9618c#diff-45ad65381ddc5fb1751709cb04576f09a0757fe43cca4cdea9e5241cc85d4a96L361-R365) you had `itemIds.Contains(x.Key)`. This means `itemIds` was enumerated `templatesDocument.Templates.Count` times, which is wasteful if the object is not already `IList<string>`.

Reply:
How about this instead?
The following code is more readable, does reduce enumeration, does not use as much memory and does not waste resources when `Remove` is not being request.
```
        if (itemIds?.Any() == true)
        {
            switch (options.BulkAction)
            {
                case ContentsBulkAction.None:
                    break;
                case ContentsBulkAction.Remove:

                    var templatesDocument = options.AdminTemplates
                        ? await _adminTemplatesManager.LoadTemplatesDocumentAsync()
                        : await _templatesManager.LoadTemplatesDocumentAsync();

                    foreach (var itemId in itemIds)
                    {
                        if (!templatesDocument.Templates.ContainsKey(itemId))
                        {
                            continue;
                        }

                        if (options.AdminTemplates)
                        {
                            await _adminTemplatesManager.RemoveTemplateAsync(itemId);

                            continue;
                        }

                        await _templatesManager.RemoveTemplateAsync(itemId);
                    }
                    await _notifier.SuccessAsync(H["Templates successfully removed."]);
                    break;
                default:
                    return BadRequest();
            }
        }
```

Reply:
The `if (itemIds?.Any() == true)` also does as an enumeration on its own. I guess it can be replaced with `if (itemIds != null)` without too much performance hit (probably even a perf gain on average as it should be very rare that an empty ID list is posted.

Also it occurred to me that this code compares keys. If we use `emplatesDocument.Templates.Keys`, then we are just looking for the union of two string sets. Linq already has an extension method for that. see [commit](https://github.com/OrchardCMS/OrchardCore/pull/16623/commits/5a167f5d00a3540f6e3d09a7e0ca12c74ff70736).

Reply:
yes `.Any()` iterates over the IEnumerable checking if there is at least one item. This is okay because it's fast since it does not iterate over every item. May be better to use `itemIds != null && itemIds.Any()` this way you don't bother loading the document or applying the union logic where there is no work to be done.